pkg/volume/util/subpath/subpath_linux_test.go

```diff
@@ -539,6 +540,64 @@ func TestCleanSubPaths(t *testing.T) {
 				return validateDirExists(baseSubdir)
 			},
 		},
+		{
+			name: "subpath-with-files",
+			prepare: func(base string) ([]mount.MountPoint, error) {
+				containerPath := filepath.Join(base, containerSubPathDirectoryName, testVol, "container1")
+				if err := os.MkdirAll(containerPath, defaultPerm); err != nil {
+					return nil, err
+				}
+
+				file0 := filepath.Join(containerPath, "0")
+				if err := ioutil.WriteFile(file0, []byte{}, defaultPerm); err != nil {
+					return nil, err
+				}
+
+				dir1 := filepath.Join(containerPath, "1")
+				if err := os.MkdirAll(filepath.Join(dir1, "my-dir-1"), defaultPerm); err != nil {
+					return nil, err
+				}
+
+				dir2 := filepath.Join(containerPath, "2")
+				if err := os.MkdirAll(filepath.Join(dir2, "my-dir-2"), defaultPerm); err != nil {
+					return nil, err
+				}
+
+				file3 := filepath.Join(containerPath, "3")
+				if err := ioutil.WriteFile(file3, []byte{}, defaultPerm); err != nil {
+					return nil, err
+				}
+
+				mounts := []mount.MountPoint{
+					{Device: "/dev/sdb", Path: file0},
+					{Device: "/dev/sdc", Path: dir1},
+					{Device: "/dev/sdd", Path: dir2},
+					{Device: "/dev/sde", Path: file3},
+				}
+				return mounts, nil
+			},
+			unmount: func(mountpath string) error {
```

Comment:
Can you explain why we need this unmount function?  The `RemoveAll` is especially scary if the test gets the path wrong.

Reply:
This is set to test with the `FakeMounter`

The unmount process expects the directory to be empty after the unmount which will then [remove the directory](https://github.com/kubernetes/kubernetes/blob/master/pkg/util/mount/mount_helper_common.go#L82)

Using the `FakeMounter` I do a clean up, to avoid it to fail in the remove

The bug itself is that without the [skipDir](https://github.com/kubernetes/kubernetes/pull/82698/files#diff-837f568dfbf46bf7db2c96e58922df54R247) it would start to try to unmount all the were inside my volume, which won't exist after the root unmount

Reply:
Where I'm confused is that there's nothing under the "0", "1", etc test directories?

Reply:
from the latest changes this is how the files look like:

```
janario@ubuntu:/tmp/clean-subpaths-subpath-with-files-825360851/volume-subpaths/vol1$ tree .
.
└── container1
    ├── 0
    ├── 1
    │   └── my-dir-1
    ├── 2
    │   └── my-dir-2
    └── 3

5 directories, 2 files

```

then the `filepath.Walk` starts with

* 0 (file): unmount function doesn't do anything because it is a file and the file gets removed [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/util/mount/mount_helper_common.go#L81)
* 1 (dir with sub dirs): unmount function will remove `my-dir-1` but not `1`
* 2 (dir with sub dirs): same as dir `1`
* 3 (file): same as file `0`

Reply:
but I think I got your point, the scenario is created from my test case, so in the unmount from my scenario should know that `my-dir-1` and `my-dir-2` will always be empty

so there is no need for a recursive remove

I've just changed it to not recursive remove

let me know what do you think

Reply:
Thanks for the explanation! I didn't realize there were directories created underneath, so this makes sense to me.

I contemplated whether or not we should just add the cleanup dir logic to the FakeMounter, but then we would need the RemoveAll to be more generic, and I'm very wary of using RemoveAll anywhere.  So I think this is fine.